sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -50,15 +51,23 @@ import org.apache.spark.unsafe.types.{ByteArray, UTF8String}
   """)
 case class Concat(children: Seq[Expression]) extends Expression with ImplicitCastInputTypes {
 
-  override def inputTypes: Seq[AbstractDataType] = Seq.fill(children.size)(StringType)
-  override def dataType: DataType = StringType
+  private lazy val isBinaryMode = children.nonEmpty && children.forall(_.dataType == BinaryType)
```

Comment:
>  If all inputs are binary, concat also outputs binary.

Is this true in Hive and others?

Reply:
will check some patterns

Reply:
`pg` and `hive` have the same;
```
postgres=# create table t1(a bytea, b bytea, c varchar, d varchar);
postgres=# create view v1 as select a || b || c || d from t1;
postgres=# \d v1
   View "public.view41_1"
  Column  | Type | Modifiers 
----------+------+-----------
 ?column? | text | 


hive> create table t1(a binary, b binary, c text, d test);
hive> create view v1 as select a || b || c || d from t1;
hive> describe v1;
_c0                 	string  
```

Reply:
Thank you for confirming it! Below is the behavior of DB2

https://www.ibm.com/support/knowledgecenter/en/SSEPGG_11.1.0/com.ibm.db2.luw.sql.ref.doc/doc/r0000781.html

Reply:
aha, thanks for the info!
I checked the db2 behaviour and I found db2 seems to have a bit different casting rule.
https://www.ibm.com/support/knowledgecenter/SSEPGG_11.1.0/com.ibm.db2.luw.sql.ref.doc/doc/r0000736.html?view=kc
IIUC, in db2, the type of concat(binary, string) is binary?

Reply:
I also checked mysql: https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_concat

recap:
hive, postgresql: concat(binary, string) => string
mysql, db2: conat(binary, string) => binary